osquery/core/database/in_memory_database.cpp

```diff
@@ -0,0 +1,147 @@
+/**
+ *  Copyright (c) 2018-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under both the Apache 2.0 license (found in the
+ *  LICENSE file in the root directory of this source tree) and the GPLv2 (found
+ *  in the COPYING file in the root directory of this source tree).
+ *  You may select, at your option, one of the above-listed licenses.
+ */
+
+#include <osquery/core/database/in_memory_database.h>
+#include <osquery/database.h>
+#include <osquery/logger.h>
+
+namespace osquery {
+
+void InMemoryDatabase::close() {
+  VLOG(1) << "Closing db... ";
+  assert(!is_open_ && "database is not open");
+  is_open_ = false;
+  destroyDB();
+}
+
+ExpectedSuccess<DatabaseError> InMemoryDatabase::destroyDB() {
+  VLOG(1) << "Destroying in memory db";
+  storage_.clear();
+  return Success();
+}
+
+ExpectedSuccess<DatabaseError> InMemoryDatabase::open() {
+  assert(is_open_ && "database is already open");
+  for (const auto& domain : kDomains) {
+    storage_[domain] = std::make_unique<InMemoryStorage<DataType>>();
+  }
+  is_open_ = true;
+  return Success();
+}
+
+Error<DatabaseError> InMemoryDatabase::domainNotFoundError(
+    const std::string& domain) {
+  return createError(DatabaseError::DomainNotFound, "Can't find domain: ")
+         << domain;
+}
```

Comment:
In would only increase complexity, right now it's looks like this 
auto storage_iter = storage_.find(domain);
  if (storage_iter == storage_.end()) {
    return domainNotFoundError(domain);
  }

and with function it will be 
auto storage = find_storage(domain); 
if (!storage) {
  return storage.takeError();
}
which is same amount of code but we have yet another function :(